### ISSUE TITLE
[Bug]: takes a long time to finish  Sourcing swe_entry.sh to set up environment variables. 

### ISSUE THREAD

**Issue by lwaekfjlk** [OpenDevin/OpenDevin#2273]
### Is there an existing issue for the same bug?

- [X] I have checked the troubleshooting document at https://opendevin.github.io/OpenDevin/modules/usage/troubleshooting
- [X] I have checked the existing issues.

### Describe the bug

I keep waiting at this stage and sometimes it can pass but sometime it does not and throw errors. Is there any way to accelerate that?

```
(base) haofeiyu@Haofeis-MacBook-Pro OpenDevin % tail -f evaluation/evaluation_outputs/outputs/swe_bench_lite/CodeActAgent/gpt-4-turbo_maxiter_30_N_v1.5/logs/instance_scikit-learn__scikit-learn-10949.log
2024-06-05 05:34:37,737 - INFO - You can debug the SSH connection by running: `ssh -v -p 52644 root@localhost` using the password 'bd9ed4ea-e332-4687-a825-3c86f7ca964d'
2024-06-05 05:34:40,226 - INFO - Connected to SSH session
2024-06-05 05:34:41,631 - INFO - Copied files from [/Users/haofeiyu/github_repo/OpenDevin/opendevin/runtime/plugins/agent_skills] to [/opendevin/plugins/agent_skills] inside sandbox.
2024-06-05 05:34:41,632 - INFO - Initializing plugin [agent_skills] by executing [/opendevin/plugins/agent_skills/setup.sh] in the sandbox.
2024-06-05 05:34:51,336 - INFO - Plugin agent_skills initialized successfully
2024-06-05 05:34:51,476 - INFO - Copied files from [/Users/haofeiyu/github_repo/OpenDevin/opendevin/runtime/plugins/jupyter] to [/opendevin/plugins/jupyter] inside sandbox.
2024-06-05 05:34:51,477 - INFO - Initializing plugin [jupyter] by executing [/opendevin/plugins/jupyter/setup.sh] in the sandbox.
2024-06-05 05:35:13,670 - INFO - Plugin jupyter initialized successfully
2024-06-05 05:35:13,914 - INFO - Sourced ~/.bashrc successfully
2024-06-05 05:35:14,516 - INFO - Sourcing swe_entry.sh to set up environment variables

```

### Current OpenDevin version

```bash
ghcr.io/opendevin/eval-swe-bench:full-v1.2.1
```

### Installation and Configuration

```bash
my config.toml is 

[core]
workspace_base="./workspace"
persist_sandbox = false
max_iterations = 100
cache_dir = "/tmp/cache"
sandbox_container_image = "ghcr.io/opendevin/sandbox:latest"
sandbox_type = "ssh"
ssh_hostname = "localhost"
sandbox_timeout = 120


# SWEBench eval specific
use_host_network = false
run_as_devin = false
enable_auto_lint = true

[eval_gpt4_1106_preview]
model = "gpt-4-1106-preview"
api_key = "xxx"
temperature = 0.0

[llm]
model="gpt-3.5-turbo"
api_key=""
embedding_model=""
```

**Comment by lwaekfjlk** [comment]
2024-06-05 05:34:41,632 - INFO - Initializing plugin [agent_skills] by executing [/opendevin/plugins/agent_skills/setup.sh] in the sandbox.
2024-06-05 05:34:51,336 - INFO - Plugin agent_skills initialized successfully
2024-06-05 05:34:51,476 - INFO - Copied files from [/Users/haofeiyu/github_repo/OpenDevin/opendevin/runtime/plugins/jupyter] to [/opendevin/plugins/jupyter] inside sandbox.
2024-06-05 05:34:51,477 - INFO - Initializing plugin [jupyter] by executing [/opendevin/plugins/jupyter/setup.sh] in the sandbox.
2024-06-05 05:35:13,670 - INFO - Plugin jupyter initialized successfully
2024-06-05 05:35:13,914 - INFO - Sourced ~/.bashrc successfully
2024-06-05 05:35:14,516 - INFO - Sourcing swe_entry.sh to set up environment variables
2024-06-05 05:45:14,619 - ERROR - Command "source /swe_util/swe_entry.sh" timed out, killing process...
2024-06-05 05:45:15,463 - INFO - exit code: -1
2024-06-05 05:45:15,464 - INFO - Command: "source /swe_util/swe_entry.sh" timed out. Sent SIGINT to the process: 
CONDA_ENV_NAME: scikit-learn__scikit-learn__0.20
Warning: 'conda-forge' already in 'channels' list, moving to the bottom
2024-06-05 09:35:41,024 - testbed_context_manager - INFO - [Testbed] Using conda path /swe_util/miniforge3
2024-06-05 09:35:41,026 - testbed_context_manager - INFO - [Testbed] Using working directory /workspace for testbed
2024-06-05 09:35:41,027 - testbed_context_manager - INFO - [Testbed] Repo scikit-learn/scikit-learn: 1 versions
2024-06-05 09:35:41,027 - testbed_context_manager - INFO - [Testbed] 	Version 0.20: 1 instances
2024-06-05 09:35:41,027 - testbed_context_manager - INFO - [Testbed] Using conda path /swe_util/miniforge3
2024-06-05 09:35:41,028 - testbed_context_manager - INFO - [Testbed] Using conda executable: /swe_util/miniforge3/bin/mamba
2024-06-05 09:35:43,576 - testbed_context_manager - INFO - [Testbed] Setting up testbed for scikit-learn__scikit-learn__0.20
2024-06-05 09:35:43,582 - testbed_context_manager - INFO - [Testbed] Repo for scikit-learn__scikit-learn version 0.20 exists: /workspace/scikit-learn__scikit-learn__0.20; skipping
2024-06-05 09:35:43,584 - testbed_context_manager - INFO - [Testbed-DEBUG]repo_path: /workspace/scikit-learn__scikit-learn__0.20
2024-06-05 09:35:43,584 - testbed_context_manager - INFO - [Testbed] Environment scikit-learn__scikit-learn__0.20 already exists; skipping
2024-06-05 09:35:45,686 - taskenv_context_manager - INFO - [scikit-learn__scikit-learn__0.20] [scikit-learn__scikit-learn-10949] Reset task environment to 3b5abf76597ce6aff76192869f92647c1b5259e7
2024-06-05 09:35:45,687 - taskenv_context_manager - INFO - [scikit-learn__scikit-learn__0.20] [scikit-learn__scikit-learn-10949] Installing with command: . /swe_util/miniforge3/etc/profile.d/conda.sh && conda activate scikit-learn__scikit-learn__0.20 && echo 'activate successful' && pip install -v --no-use-pep517 --no-build-isolation -e .
Traceback (most recent call last):
  File "/swe_util/OD-SWE-bench/swebench/harness/reset_swe_env.py", line 164, in <module>
    main(**vars(args))
  File "/swe_util/OD-SWE-bench/swebench/harness/reset_swe_env.py", line 142, in main
    if not setup_swe_env(task_set, apply_test_patch=False, run_gold_tests=False):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/swe_util/OD-SWE-bench/swebench/harness/reset_swe_env.py", line 63, in setup_swe_env
    if not tcm.run_install_task(task_instance):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/swe_util/OD-SWE-bench/swebench/harness/context_manager.py", line 625, in run_install_task
    out_install = self.exec(cmd_install, timeout=self.timeout, shell=True, env=None)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/swe_util/OD-SWE-bench/swebench/harness/context_manager.py", line 55, in __call__
    output = subprocess.run(cmd, **combined_args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/swe_util/miniforge3/envs/swe-bench-eval/lib/python3.11/subprocess.py", line 550, in run
    stdout, stderr = process.communicate(input, timeout=timeout)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/swe_util/miniforge3/envs/swe-bench-eval/lib/python3.11/subprocess.py", line 1209, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/swe_util/miniforge3/envs/swe-bench-eval/lib/python3.11/subprocess.py", line 2108, in _communicate
    ready = selector.select(timeout)
            ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/swe_util/miniforge3/envs/swe-bench-eval/lib/python3.11/selectors.py", line 415, in select
    fd_event_list = self._selector.poll(timeout)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt

potential situation I face

**Comment by SmartManoj** [comment]
Increase the timeout.

**Comment by lwaekfjlk** [comment]
is timeout a paramter in the config?

**Comment by SmartManoj** [comment]
that's hardcoded 

https://github.com/OpenDevin/OpenDevin/blob/8189115acf9ad17c0d76359a3d5787f19814fb8d/evaluation/swe_bench/swe_env_box.py#L57

**Comment by tobitege** [comment]
Did you follow the instructions here?
https://github.com/OpenDevin/OpenDevin/blob/main/evaluation/swe_bench/BUILD_TESTBED_AND_ENV.md

**Comment by lwaekfjlk** [comment]
@tobitege this is non-docker prepare, right? I am using docker to run.

**Comment by SmartManoj** [comment]
https://github.com/OpenDevin/OpenDevin/blob/main/evaluation/swe_bench/BUILD_TESTBED_AND_ENV.md#build-the-pre-build-conda-env-and-testbed-into-one-docker-image

**Comment by tobitege** [comment]
I read the bottom docker commands are only meant for the maintainers to create containers in the cloud.

But I think @lwaekfjlk you should be able to run this inside your docker image:
`./evaluation/swe_bench/scripts/setup/prepare_swe_utils.sh`

**Comment by lwaekfjlk** [comment]
yes, I am able to run that

**Comment by tobitege** [comment]
> yes, I am able to run that

Ok, good. If you run this manually, it should finish without a timeout. And any subsequent call to the evaluation should run faster since the env is (should) already be prepared.

That's at least my best educated guess, I haven't run it myself yet. 😬

**Comment by lwaekfjlk** [comment]
I think my issue still exists after doing that. This issue is similar with #2063

**Comment by SmartManoj** [comment]
> I read the bottom docker commands

Where?

**Comment by tobitege** [comment]
at the bottom of the readme you posted

**Comment by SmartManoj** [comment]
[Last word](https://github.com/OpenDevin/OpenDevin/blob/main/evaluation/swe_bench/BUILD_TESTBED_AND_ENV.md#build-the-pre-build-conda-env-and-testbed-into-one-docker-image:~:text=push%20ghcr.io/opendevin/eval%2Dswe%2Dbench%3A-,full%2Dv1.2.1)

**Comment by tobitege** [comment]
Apologies, I did not correctly understand. So I asked for help: 😁 
`
Let's break down each of the commands:

1. **`pushd evaluation/swe_bench`**:
   - This command changes the current directory to `evaluation/swe_bench` and saves the current directory on the directory stack. The `pushd` command is useful for navigating directories while keeping track of where you came from. You can return to the original directory with the `popd` command.

2. **`docker build -t ghcr.io/opendevin/eval-swe-bench:full-v1.2.1 -f ./scripts/docker/Dockerfile.full.v1.1 .`**:
   - This command builds a Docker image using the Dockerfile located at `./scripts/docker/Dockerfile.full.v1.1`.
   - The `-t` flag tags the resulting image with the name `ghcr.io/opendevin/eval-swe-bench:full-v1.2.1`.
   - The `.` at the end specifies the build context, which is the current directory (`evaluation/swe_bench`).

3. **`docker push ghcr.io/opendevin/eval-swe-bench:full-v1.2.1`**:
   - This command pushes the Docker image `ghcr.io/opendevin/eval-swe-bench:full-v1.2.1` to the GitHub Container Registry.
   - This makes the image available for others to pull and use from the GitHub Container Registry.

In summary, these commands change to the `evaluation/swe_bench` directory, build a Docker image from a specified Dockerfile, and then push the resulting image to the GitHub Container Registry.`

**Comment by tobitege** [comment]
@SmartManoj so is running below **INSIDE** the container the right way?
`sudo source /swe_util/swe_entry.sh`

**Comment by SmartManoj** [comment]
the setup will skip if it is already installed?

**Comment by lwaekfjlk** [comment]
@SmartManoj My understanding is that that is for the developer's to reproduce the docker environment. However, I have already pulled the docker and I do not need to do anything. However, there is something wrong with my docker (even though I pull to make sure it is the latest), so that I cannot `swe_env_box.py:53 - Sourcing swe_entry.sh to set up environment variables` very fluently.

**Comment by tobitege** [comment]
The individual packages are skipped if already installed, so should run faster next time.

**Comment by lwaekfjlk** [comment]
@tobitege 

I am able to run the swe_entry.sh


root@bab35b1c4715:/swe_util# source swe_entry.sh 
CONDA_ENV_NAME: scikit-learn__scikit-learn__0.22
Warning: 'conda-forge' already in 'channels' list, moving to the bottom
2024-06-05 11:26:27,435 - testbed_context_manager - INFO - [Testbed] Using conda path /swe_util/miniforge3
2024-06-05 11:26:27,437 - testbed_context_manager - INFO - [Testbed] Using working directory /workspace for testbed
2024-06-05 11:26:27,438 - testbed_context_manager - INFO - [Testbed] Repo scikit-learn/scikit-learn: 1 versions
2024-06-05 11:26:27,438 - testbed_context_manager - INFO - [Testbed] 	Version 0.22: 1 instances
2024-06-05 11:26:27,439 - testbed_context_manager - INFO - [Testbed] Using conda path /swe_util/miniforge3
2024-06-05 11:26:27,439 - testbed_context_manager - INFO - [Testbed] Using conda executable: /swe_util/miniforge3/bin/mamba
2024-06-05 11:26:28,964 - testbed_context_manager - INFO - [Testbed] Setting up testbed for scikit-learn__scikit-learn__0.22
2024-06-05 11:26:28,965 - testbed_context_manager - INFO - [Testbed] Repo for scikit-learn__scikit-learn version 0.22 exists: /workspace/scikit-learn__scikit-learn__0.22; skipping
2024-06-05 11:26:28,965 - testbed_context_manager - INFO - [Testbed-DEBUG]repo_path: /workspace/scikit-learn__scikit-learn__0.22
2024-06-05 11:26:28,965 - testbed_context_manager - INFO - [Testbed] Environment scikit-learn__scikit-learn__0.22 already exists; skipping
2024-06-05 11:26:30,614 - taskenv_context_manager - INFO - [scikit-learn__scikit-learn__0.22] [scikit-learn__scikit-learn-13779] Reset task environment to b34751b7ed02b2cfcc36037fb729d4360480a299
2024-06-05 11:26:30,615 - taskenv_context_manager - INFO - [scikit-learn__scikit-learn__0.22] [scikit-learn__scikit-learn-13779] Installing with command: . /swe_util/miniforge3/etc/profile.d/conda.sh && conda activate scikit-learn__scikit-learn__0.22 && echo 'activate successful' && pip install -v --no-use-pep517 --no-build-isolation -e .

but it also stuck and have no output for a long time after what the above has been printed

**Comment by tobitege** [comment]
I can't really help with this as that's not my domain.

The only thing I could suggest by now is to edit the file `evaluation/swe_bench/swe_env_box.py`, then replace in line 55 the value 600 with e.g. 3600, save the file and run the evaluation again.
I'd assume that your installation is not stuck, but a huge amount of stuff that needs to be downloaded and installed, probably several GB's.

**Comment by tobitege** [comment]
Hello @lwaekfjlk 
please have a look at this related issue with a potential fix:
 https://github.com/OpenDevin/OpenDevin/issues/2285#issuecomment-2151344986

Before running the eval, run this in terminal:
`export PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring`

@SmartManoj what do you think: should this maybe added to Makefile if the env var doesn't exist?

**Comment by SmartManoj** [comment]
> related issue

how it is?

**Comment by tobitege** [comment]
ahh, nvm, I need more coffee, misread logfile... 🤕 ☕

**Comment by zkx06111** [comment]
same here. Sometimes when it's stuck I try to restart docker engine and it starts working again. Not sure why.

**Comment by lwaekfjlk** [comment]
I solve this issue by `make build` without conda or venv and launch the job again. But I actually don't understand the reason behind it. @tobitege @SmartManoj  Any thought on that? I think Conda create some communication error when launching the docker?

**Comment by SmartManoj** [comment]
> (base) haofeiyu@Haofeis-MacBook-Pro OpenDevin

Because Pexpect expects a default prompt string.
